src/utils/timeline.h

```diff
@@ -171,23 +174,77 @@ class TimeLine {
         close(lockFileDescriptor);
     }
 
+    inline void startTimeLineEvent(const std::string &name){
+        tag_name = name;
+        pid = getpid();
+        pid = pthread_self();
+        trace_event["ph"] = "X";
+        trace_event["cat"] = "cat";
+        trace_event["name"] = tag_name.c_str();
+        trace_event["pid"] = pid;
+        trace_event["tid"] = tid;
+        during_time = -1;
+        start = std::chrono::high_resolution_clock::now();
+        end = start;
+    };
+
+    static void strip(std::string &word){
+        std::string whitespace("\n\t ");
+        std::size_t found = word.find_last_not_of(whitespace);
+        if (found!=std::string::npos) {
+            word.erase(found+1);
+            found = word.find_first_not_of(whitespace);
+            if (found!=std::string::npos)
+                word.erase(0, found);
+        }
+        else
+            word.clear();
+    }
+
+    static void init_tag_whitelist() {
+        char *value = getenv("XFT_TIMELINE_WHITELIST");
+        tag_whitelist.clear();
+        if (value) {
+            std::string env(value);
+            size_t start = 0, end;
+            while ((end = env.find(',', start)) != std::string::npos) {
+                auto event = std::string(env.substr(start, end - start));
+                strip(event);
+                if(!event.empty())
+                    tag_whitelist[event] = 1;
+                start = end + 1;
+            }
+            auto event = std::string(env.substr((start)));
+            strip(event);
+
+            if(!event.empty())
+                tag_whitelist[event] = 1;
+            if (!tag_whitelist.empty())
+                having_whitelist = true;
+        }
+    }
+
     std::string tag_name;
     int64_t pid;
     int64_t tid;
     std::chrono::high_resolution_clock::time_point start;
     std::chrono::high_resolution_clock::time_point end;
-    int64_t start_timestap, during_time;
+    int64_t start_timestamp, during_time;
     Json::Value trace_event;
+    static inline std::vector<Json::Value> pool{};
+    static inline std::unordered_map<std::string, int32_t> tag_whitelist{};
```

Comment:
Sorry, I don't know why to use std::unordered_map structure. vector?

Reply:
Got. for avoiding duplicate lookups

Reply:
This is for the case of a large 'whitelist', looking for an event name in unordered_map will be faster than vector list. A unordered_map is a "hash_map", which means that searching it is O(1).